x/trigger/keeper/trigger_dispatcher.go

```diff
@@ -0,0 +1,117 @@
+package keeper
+
+import (
+	"fmt"
+
+	"github.com/cosmos/cosmos-sdk/baseapp"
+	types "github.com/cosmos/cosmos-sdk/codec/types"
+	storetypes "github.com/cosmos/cosmos-sdk/store/types"
+	sdk "github.com/cosmos/cosmos-sdk/types"
+	sdktx "github.com/cosmos/cosmos-sdk/types/tx"
+)
+
+const (
+	MaximumActions  uint64 = 5
+	MaximumQueueGas uint64 = 2000000
+)
+
+// ProcessTriggers Reads triggers from queues and attempts to run them.
+func (k Keeper) ProcessTriggers(ctx sdk.Context) {
+	var actionsProcessed uint64
+	var gasConsumed uint64
+
+	for !k.QueueIsEmpty(ctx) && actionsProcessed < MaximumActions {
+		item := k.QueuePeek(ctx)
+		triggerID := item.GetTrigger().Id
+		gasLimit := k.GetGasLimit(ctx, triggerID)
+
+		if gasLimit+gasConsumed > MaximumQueueGas {
+			return
+		}
+		actionsProcessed++
+		gasConsumed += gasLimit
+
+		k.Dequeue(ctx)
+		k.RemoveGasLimit(ctx, triggerID)
+
+		actions := item.GetTrigger().Actions
+		k.runActions(ctx, gasLimit, actions)
+	}
+}
+
+// RunActions Runs all the actions and constrains them by gasLimit.
+func (k Keeper) runActions(ctx sdk.Context, gasLimit uint64, actions []*types.Any) {
+	cacheCtx, flush := ctx.CacheContext()
+	gasMeter := sdk.NewGasMeter(gasLimit)
+	cacheCtx = cacheCtx.WithGasMeter(gasMeter)
+
+	msgs, err := sdktx.GetMsgs(actions, "RunActions")
+	if err != nil {
+		k.Logger(ctx).Error(
+			"GetMsgs",
+			"actions", actions,
+			"error", err,
+		)
+		return
+	}
+	results, err := k.handleMsgs(cacheCtx, msgs)
+	if err != nil {
```

Comment:
Regardless of error or success the gas used in `handleMsgs` must be consumed against the block gas meter.
```
https://github.com/cosmos/cosmos-sdk/blob/e7dd5e0bdc8aabadb0d4ae13bdb143f92c1cdf4c/baseapp/baseapp.go#L640-L658

Reply:
There is a subtle difference in what I was referring to ... you need to consume gas from the `BlockGasMeter` which is on the `ctx` and is not the same as your local transaction gas meter.